### PR TITLE
`FileCache`: detect modified source files

### LINKED ISSUE #1602: Make `FileCache` able to detect updated remote files
Sometimes corrupt or old files are updated/replaced on the remote file system. If a file is already cached on the local machine, the `FileCache` will not detect this situation and happily use the locally cached variant. I think we can improve on this by moving the cache-keepalive `mtime`-logic to a proxy file, e.g. `original-file.name.keepalive` and comparing the `mtime` of the actual data files.

### PR DESCRIPTION
**Pull request by NeoLegends**
...by storing the `mtime`s in dedicated keepalive files next to the cached files themselves.

Closes #1602 

@Icemole Do you want to test this in your setup?

### CODE REVIEW

**Comment by michelwi** [comment]
But this would not keep the nodes tmp cleanup scripts from deleting the data files.. would it?

**Comment by NeoLegends** [comment]
No, this is not something the `FileCache` deals with anyway, though. What do the temp reaper scripts look at to determine if a file is to be cleaned up?

File: returnn/util/file_cache.py [hunk 1]
```diff
@@ -337,8 +363,12 @@ def _check_existing_copied_file_maybe_cleanup(src_filename: str, dst_filename: s
             return False
         src_stat = os.stat(src_filename)
         dst_stat = os.stat(dst_filename)
-        if src_stat.st_size != dst_stat.st_size:
+        if src_stat.st_size != dst_stat.st_size or src_stat.st_mtime > dst_stat.st_mtime:
             os.remove(dst_filename)
+            try:
+                os.remove(FileCache._get_keepalive_filename(dst_filename))
```
Comment (Icemole) [inline returnn/util/file_cache.py:369]: In which situations can a keepalive file not be along with the file itself?

File: returnn/util/file_cache.py [hunk 2]
```diff
@@ -337,8 +363,12 @@ def _check_existing_copied_file_maybe_cleanup(src_filename: str, dst_filename: s
...
+            except FileNotFoundError:
+                pass
```
Comment (Icemole) [inline returnn/util/file_cache.py:371]: Above there's the following code when the keepalive file can't be found:
```
print(
    f"FileCache: Ignoring error file removing keepalive file of {fn}: {type(exc).__name__}: {exc}"
)
```

I'd suggest to uniformize the calls by either setting the print here (recommended), abstracting this into a function (also recommended, as you prefer) or getting rid of the original print and adding a `pass` instead.

Requesting changes for this.

File: returnn/util/file_cache.py [hunk 3]
```diff
@@ -35,16 +35,14 @@ class FileCache:
 
     Some assumptions we depend on:
 
-    - When a cached file is available, and its size matches the original file,
+    - When a cached file is available,
+      and its size matches the original file
+      and its mtime is not older than the remove file,
```
Comment (Icemole) [inline returnn/util/file_cache.py:40]:
```suggestion
      and its mtime is not older than the original file,
```

**Review by Icemole — CHANGES_REQUESTED** [review]

**Comment by Icemole** [comment]
Yes, I can test this with a small setup I had lying around. Let me know once you fix all tests via slack so that I test the final functionality.

**Comment by albertz** [comment]
> ...by storing the `mtime`s in dedicated keepalive files next to the cached files themselves.

As discussed in #1602, wouldn't it make more sense the other way around?

I think it would cause some other unexpected problems. E.g. the [tmpreaper](https://manpages.ubuntu.com/manpages/trusty/man8/tmpreaper.8.html) would cleanup the files (it looks at mtime).

**Comment by NeoLegends** [comment]
I like this idea.

> No, this is not something the FileCache deals with anyway, though

... and I stand corrected. Apparently since we update the mtimes of the files that are alive, they are indeed protected from the tmpreaper.

File: returnn/util/file_cache.py [hunk 4]
```diff
@@ -397,6 +436,29 @@ class CachedFile:
...
+
+    We currently save the last known mtime of the source file.
+    """
+
+    mtime_ns: int
+
+    def save(self, filename: str):
+        """Saves this instance to `filename`."""
+        with open(filename, "wt") as f:
+            f.write(json.dumps(dataclasses.asdict(self)))
```
Comment (NeoLegends) [inline returnn/util/file_cache.py:452]: JSON and not pickle, btw, because it is human readable.

Comment (albertz) [reply returnn/util/file_cache.py:452]: Note, in most other cases, we store as Python string (repr), and then use eval (or ast eval). But Json is also fine.

**Review by Icemole — APPROVED** [review]
I've started a small training with this patch and the dev frame error rates are practically equivalent to the original ones. I'm approving.

**Review by michelwi — APPROVED** [review]

**Pull request merged**